### PR TITLE
lxd: Retry if the server isn't ready

### PR DESCRIPTION
**Pull request by holmanb**
```
lxd: Retry if the server isn't ready
```

## Additional Context
```
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/cloudinit/sources/__init__.py", line 946, in find_source
    if s.update_metadata_if_supported(
  File "/usr/lib/python3/dist-packages/cloudinit/sources/__init__.py", line 828, in update_metadata_if_supported
    result = self.get_data()
  File "/usr/lib/python3/dist-packages/cloudinit/sources/__init__.py", line 373, in get_data
    return_value = self._get_data()
  File "/usr/lib/python3/dist-packages/cloudinit/sources/DataSourceLXD.py", line 193, in _get_data
    self._crawled_metadata = util.log_time(
  File "/usr/lib/python3/dist-packages/cloudinit/util.py", line 2680, in log_time
    ret = func(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/cloudinit/sources/DataSourceLXD.py", line 428, in read_metadata
    return _MetaDataReader(api_version=api_version)(
  File "/usr/lib/python3/dist-packages/cloudinit/sources/DataSourceLXD.py", line 384, in __call__
    md["meta-data"] = _do_request(session, md_route).text
  File "/usr/lib/python3/dist-packages/cloudinit/sources/DataSourceLXD.py", line 297, in _do_request
    raise sources.InvalidMetaDataException(
cloudinit.sources.InvalidMetaDataException: Invalid HTTP response [500] from http://lxd/1.0/meta-data: internal server error
```

## Test Logs
```
2023-02-17 23:25:24,815 - __init__.py[DEBUG]: Update datasource metadata and network config due to events: boot-new-instance
2023-02-17 23:25:25,039 - DataSourceLXD.py[WARNING]: [GET] [HTTP:500] http://lxd/1.0/meta-data, retrying
2023-02-17 23:25:25,143 - DataSourceLXD.py[WARNING]: [GET] [HTTP:500] http://lxd/1.0/meta-data, retrying
2023-02-17 23:25:25,179 - DataSourceLXD.py[DEBUG]: [GET] [HTTP:200] http://lxd/1.0/meta-data
2023-02-17 23:25:25,229 - DataSourceLXD.py[DEBUG]: [GET] [HTTP:200] http://lxd/1.0/config
2023-02-17 23:25:25,269 - DataSourceLXD.py[DEBUG]: [GET] [HTTP:200] http://lxd/1.0/config/cloud-init.user-data
2023-02-17 23:25:25,359 - DataSourceLXD.py[DEBUG]: [GET] [HTTP:200] http://lxd/1.0/devices
```

### CODE REVIEW

File: cloudinit/sources/DataSourceLXD.py [hunk 1]
```diff
@@ -299,7 +300,17 @@ def _get_json_response(
 def _do_request(
     session: requests.Session, url: str, do_raise: bool = True
 ) -> requests.Response:
-    response = session.get(url)
+    for _ in range(30):
```
Comment (blackboxsw) [inline cloudinit/sources/DataSourceLXD.py:303]: Is it worth keeping the var around to report how many retries remain?
```suggestion
    for tries in range(30):
    for retries in range(29, 0, -1):
    ....
    LOG.warning(
        "[GET] [HTTP:%d] %s, retrying %d more time(s)", response.status_code, url, retries
```

**Review by blackboxsw — COMMENTED** [review]
Changeset looks fine, I'd like a breadcrumb counter in logs so we can track how many retries were perform easily. I realize a wc -l would do this. so no biggie. Can we unit test this ?

**Comment by holmanb** [comment]
Test logs from latest iteration:
```
2023-02-18 00:02:01,080 - DataSourceLXD.py[WARNING]: [GET] [HTTP:500] http://lxd/1.0/meta-data, retrying 29 more time(s)
2023-02-18 00:02:01,186 - DataSourceLXD.py[WARNING]: [GET] [HTTP:500] http://lxd/1.0/meta-data, retrying 28 more time(s)
2023-02-18 00:02:01,290 - DataSourceLXD.py[WARNING]: [GET] [HTTP:500] http://lxd/1.0/meta-data, retrying 27 more time(s)
2023-02-18 00:02:01,412 - DataSourceLXD.py[WARNING]: [GET] [HTTP:500] http://lxd/1.0/meta-data, retrying 26 more time(s)
2023-02-18 00:02:01,516 - DataSourceLXD.py[WARNING]: [GET] [HTTP:500] http://lxd/1.0/meta-data, retrying 25 more time(s)
2023-02-18 00:02:01,576 - DataSourceLXD.py[DEBUG]: [GET] [HTTP:200] http://lxd/1.0/meta-data
2023-02-18 00:02:01,606 - DataSourceLXD.py[DEBUG]: [GET] [HTTP:200] http://lxd/1.0/config
2023-02-18 00:02:01,629 - DataSourceLXD.py[DEBUG]: [GET] [HTTP:200] http://lxd/1.0/config/cloud-init.user-data
2023-02-18 00:02:01,661 - DataSourceLXD.py[DEBUG]: [GET] [HTTP:200] http://lxd/1.0/devices
```

File: tests/unittests/sources/test_lxd.py [hunk 2]
```diff
@@ -697,5 +697,42 @@ def test_read_metadata_keys(
...
+                        for code in return_codes
+                    ]
+                )
+            )
+            lxd._do_request(m, "http://agua/")
 
-# vi: ts=4 expandtab
+            # assert that 30 iterations or the first 200 code is the final
+            # attempt, whichever comes first
+            assert min(len(return_codes), 30) == m.get.call_count
```
Comment (blackboxsw) [inline tests/unittests/sources/test_lxd.py:738]: Nice and thorough test. :)

Comment (blackboxsw) [inline tests/unittests/sources/test_lxd.py:738]: Let's validate we get the right status_code too on 'failure/timeout'

```diff

diff --git a/tests/unittests/sources/test_lxd.py b/tests/unittests/sources/test_lxd.py
index dd5f90b4c..efc248838 100644
--- a/tests/unittests/sources/test_lxd.py
+++ b/tests/unittests/sources/test_lxd.py
@@ -731,8 +731,12 @@ class TestReadMetadata:
                     ]
                 )
             )
-            lxd._do_request(m, "http://agua/")
+            resp = lxd._do_request(m, "http://agua/")
 
             # assert that 30 iterations or the first 200 code is the final
             # attempt, whichever comes first
             assert min(len(return_codes), 30) == m.get.call_count
+            if len(return_codes) < 31:
+                assert 200 == resp.status_code
+            else:
+                assert 500 == resp.status_code
```

**Review by blackboxsw — APPROVED** [review]
minor diff in unit test and +1. If github behaves let's land it

**Pull request merged**